src/HIVCohortFilter/HIVCohortCases.jsx

```diff
@@ -0,0 +1,1205 @@
+      );
+  }
+
+  makeSubjectToVisitMap(followUps) { // eslint-disable-line class-methods-use-this
+    // Convert to dictionary: { subject_id -> [ array of visits sorted by visit_date ] }
+    const subjectToVisitMap = {};
+    for (let i = 0; i < followUps.length; i += 1) {
+      const subjectId = followUps[i].subject_id;
+      if (subjectId in subjectToVisitMap) {
+        subjectToVisitMap[subjectId].push(followUps[i]);
+      } else {
+        subjectToVisitMap[subjectId] = [followUps[i]];
+      }
+    }
+
+    // Sort each patient's visits by visit_number
+    Object.keys(subjectToVisitMap).forEach((key) => {
+      const subjectVisits = subjectToVisitMap[key];
+      if (subjectVisits.length > 1) {
+        subjectVisits.sort((a, b) => {
+          if (a.visit_number > b.visit_number) {
+            return 1;
+          }
+          return ((b.visit_number > a.visit_number) ? -1 : 0);
+        });
+        subjectToVisitMap[key] = subjectVisits;
+      }
+    });
+
+    return subjectToVisitMap;
+  }
+
+  updateFilters(event) {
+    event.preventDefault();
+    this.setState({ inLoadingState: true });
+    this.updateSubjectClassifications();
+  }
+
+  checkReadyToCalculate() {
+    // Overridden by LTNP case
+    const viralLoadFromUser = this.viralLoadInputRef.current.valueAsNumber;
+    const numConsecutiveMonthsFromUser = this.numConsecutiveMonthsInputRef.current.valueAsNumber;
+    this.setState({
+      viralLoadFromUser: viralLoadFromUser > 0 ? viralLoadFromUser : undefined,
+      numConsecutiveMonthsFromUser: numConsecutiveMonthsFromUser > 0
+        ? numConsecutiveMonthsFromUser : undefined,
+      isReadyToCalculate: (viralLoadFromUser > 0 && numConsecutiveMonthsFromUser > 0),
+      resultAlreadyCalculated: false,
+    });
+  }
+
+  async getBucketByKey(bucketKey) {
+    // Overridden by PTC case
+    // Returns map of subjects who have never received HAART treatment
+    const resList = await Promise.all([
+      this.getBucketByKeyWithHAARTVAR(bucketKey, true),
+      this.getBucketByKeyWithHAARTVAR(bucketKey, false),
+    ]);
+
+    const subjectsWithAtLeast1Haart = resList[0];
+    const subjectsWithAtLeast1NonHaart = resList[1];
+    const subjectsWithNoHaartTreatments = subjectsWithAtLeast1NonHaart.filter(
+      x => !subjectsWithAtLeast1Haart.map(y => y.key).includes(x.key)
+    );
+
+    // Transform to map
+    const resultMap = {};
+    for (let i = 0; i < subjectsWithNoHaartTreatments.length; i += 1) {
+      resultMap[subjectsWithNoHaartTreatments[i].key] = subjectsWithNoHaartTreatments[i].doc_count
+    }
+    return resultMap;
+  }
+
+  downloadControl() {
+    // Overridden by LTNP
+    const fileName = `control-cohort-vload-${this.state.viralLoadFromUser.toString()
+    }-months-${this.state.numConsecutiveMonthsFromUser.toString()}.json`;
+
+    const blob = this.makeCohortJSONFile(this.state.subjectControl);
+    FileSaver.saveAs(blob, fileName);
+  }
+}
+
+class PTCCase extends HIVCohortFilterCase {
```

Comment:
Is it possible `PTCCase`, `ECCase`, `LTNPCase` to separate files? this files over 1k+ lines 😛

Reply:
✓